src/main/java/com/cloudant/client/api/Database.java

```diff
@@ -1029,135 +1029,67 @@ public void ensureFullCommit() {
+            fieldObject.addProperty(fields[i].getName(), fields[i].getOrder().toString());
+            fieldsArray.add(fieldObject);
         }
+        JsonObject arrayOfFields = new JsonObject();
+        arrayOfFields.add("fields", fieldsArray);
+        indexObject.add("index", arrayOfFields);
 
-        return json + "] }}";
+        return indexObject;
     }
 
-    private String getFindByIndexBody(String selectorJson,
+    private JsonObject getFindByIndexBody(String selectorJson,
                                       FindByIndexOptions options) {
 
-        StringBuilder rf = null;
+        JsonObject fieldsObject = new JsonObject();
         if (options.getFields().size() > 0) {
-            rf = new StringBuilder("\"fields\": [");
-            int i = 0;
-            for (String s : options.getFields()) {
-                if (i > 0) {
-                    rf.append(",");
-                }
-                rf.append("\"").append(s).append("\"");
-                i++;
-            }
-            rf.append("]");
+            JsonElement fieldsArray = getGson().toJsonTree(options.getFields());
+            fieldsObject.add("fields", fieldsArray.getAsJsonArray());
         }
 
-        StringBuilder so = null;
         if (options.getSort().size() > 0) {
-            so = new StringBuilder("\"sort\": [");
-            int i = 0;
-            for (IndexField idxfld : options.getSort()) {
-                if (i > 0) {
-                    so.append(",");
-                }
-                so.append("{\"")
-                        .append(idxfld.getName())
-                        .append("\": \"")
-                        .append(idxfld.getOrder())
-                        .append("\"}");
-                i++;
-            }
-            so.append("]");
+            JsonElement sortArray = getGson().toJsonTree(options.getSort());
+            fieldsObject.add("sort", sortArray.getAsJsonArray());
         }
 
-        //parse and find if valid json issue #28
-        boolean isObject = true;
-        try {
-            client.getGson().fromJson(selectorJson, JsonObject.class);
-        } catch (JsonParseException e) {
-            isObject = false;
-        }
-
-        if (!isObject) {
-            // needs to start with selector
-            if (!(selectorJson.trim().startsWith("\"selector\""))) {
-                throw new JsonParseException("selectorJson should be valid json or like " +
-                        "\"selector\": {...} ");
-            }
-        }
 
-        StringBuilder finalbody = new StringBuilder();
-        if (isObject) {
-            finalbody.append("{\"selector\": ")
-                    .append(selectorJson);
-        } else {
-            //old support
-            finalbody.append("{" + selectorJson);
-        }
+        JsonObject indexObject = new JsonObject();
+        //Add selector json object
+        indexObject.addProperty("selector", selectorJson);
```

Comment:
Isn't the index object going to end up with: `"selector" : "\"selector\" : {...}` because the string`selectorJson` at this point includes `"selector" : ` at the beginning.
If you run the cloudantServiceTest you'll see what I mean.